src/scores/probability/crps_impl.py

```diff
@@ -869,6 +872,15 @@ def crps_for_ensemble(
     fcst_obs_term = abs(fcst - obs).mean(dim=ensemble_member_dim)
     result = fcst_obs_term - fcst_spread_term
 
+    if decomposition:
+        mask = ~np.isnan(fcst) * ~np.isnan(obs)  # create mask so that we can preserve NaNs
+        under_penalty = (obs - fcst).where(fcst < obs, 0).where(mask).mean(dim=ensemble_member_dim)
+        over_penalty = (fcst - obs).where(fcst > obs, 0).where(mask).mean(dim=ensemble_member_dim)
```

Comment:
The mask handling here is different to the calculation without decomposition. I think it might be possible to create an artificial case where the fcst_spread_term is different depending whether or not decomposition is requested. I am unsure whether either approach is 'more correct'. I will try to construct a test case which shows this (as perhaps I am wrong), and if it bears out I would suggest things do need to be consistent.

Reply:
Okay, I'm happy for you to construct a test case and try and produce an inconsistency.

One thing that may note be clear immediately:
1. No decomposition. NaNs are broadcast with `result = fcst_obs_term - fcst_spread_term`. The `fcst_spread_term` is never returned, so we only need the NaN broadcasting to flow through to `result`
2. Decomposition. NaNs are broadcast with `fcst_spread_term = fcst_spread_term.where(~np.isnan(fcst_obs_term))`

Reply:
Okay, so I'm starting to review this now. The "result" term (as I think of it, the primary term) is always identical in the code as it's not recalculated when decomposition is set to True.

The different masking applies to the under and over penalties, but also to the fcst_spread_term. I modified the function to return the fcst_spread_term, and it is indeed sometimes different. What is less clear is whether it's possible for this difference to result in any mathematical differences. The difference wasn't in specific values, but in whether things were NaNs or not.

I tried modifying the block under "if decomposition" to recalculate the 'result' from the differently-masked fcst_spread_term, then return that recalculated result and see if I could find any difference that way. I did not find any difference. 

That's good news, but it could just be I didn't quite hit the edge case. In order to even be able to test for it, I had to modify the function to (a) recalculate the result from the masked fcst_spread_term and (b) return more things. This make it hard to put in place automated tests.

I am presuming the apply_weights function is skipping the NaNs in getting at the final result, meaning all said and done the answer is the same. 

If it always did the recalculation, then perhaps more assurance could be gained by adding further test coverage, but it also seems somewhat needless given that the manual testing revealed no issues.

Perhaps you could take a look at the code in light of these comments, and consider whether you think there's room for an issue, and I will happily go with your assessment after considering things. If you think it's worthwhile, feel free to pull in another reviewer. If it would be helpful, I can share my test code in a branch.

Reply:
Thanks for looking into it. I am comfortable with it as is since the total result should always be identical in both situations. The fcst_spread term will differ as you mentioned, but we don't return it when `decomposition=False` and masking it to make it identical in both steps would lead to an unnecessary computation step (since the matching of NaNs is handled in `result = fcst_obs_term - fcst_spread_term`)

> I am presuming the apply_weights function is skipping the NaNs in getting at the final result, meaning all said and done the answer is the same.

Yes, the apply_weights function handles NaNs

I'm happy to add more unit tests if there is a specific case that you can think of, but the tests already check that NaN handling is done.

Reply:
Agreed. Okay, thanks, always good to work through things to make sure.